tests/unit/integration/vulnerabilities/osv/test_package.py

```diff
@@ -0,0 +1,466 @@
+                    "-----END PUBLIC KEY-----\n",
+                    "is_current": True,
+                }
+            ]
+        }
+        response = pretend.stub(
+            json=lambda: meta_payload, raise_for_status=lambda: None
+        )
+        session = pretend.stub(get=pretend.call_recorder(lambda *a, **k: response))
+        metrics = pretend.stub(increment=pretend.call_recorder(lambda str: None))
+
+        vuln_report_verifier = osv.VulnerabilityReportVerifier(
+            public_keys_api_url="http://foo",
+            session=session,
+            metrics=metrics,
+            public_keys_cache=pretend.stub(),
+        )
+        assert vuln_report_verifier.retrieve_public_key_payload() == meta_payload
+        assert session.get.calls == [
+            pretend.call(
+                "http://foo",
+            )
+        ]
+
+    def test_get_cached_public_key_cache_hit(self):
+        metrics = pretend.stub()
+        session = pretend.stub()
+        cache = integrations.PublicKeysCache(cache_time=12)
+        cache_value = pretend.stub()
+        cache.set(now=time.time(), value=cache_value)
+
+        vuln_report_verifier = osv.VulnerabilityReportVerifier(
+            public_keys_api_url="http://foo",
+            session=session,
+            metrics=metrics,
+            public_keys_cache=cache,
+        )
+
+        assert vuln_report_verifier._get_cached_public_keys() is cache_value
+
+    def test_get_cached_public_key_cache_miss_no_cache(self):
+        metrics = pretend.stub()
+        session = pretend.stub()
+        cache = integrations.PublicKeysCache(cache_time=12)
+
+        vuln_report_verifier = osv.VulnerabilityReportVerifier(
+            public_keys_api_url="http://foo",
+            session=session,
+            metrics=metrics,
+            public_keys_cache=cache,
+        )
+
+        with pytest.raises(integrations.CacheMiss):
+            vuln_report_verifier._get_cached_public_keys()
+
+    def test_retrieve_public_key_payload_http_error(self):
+        response = pretend.stub(
+            status_code=418,
+            text="I'm a teapot",
+            raise_for_status=pretend.raiser(requests.HTTPError),
+        )
+        session = pretend.stub(
+            get=lambda *a, **k: response,
+        )
+        vuln_report_verifier = osv.VulnerabilityReportVerifier(
+            public_keys_api_url="http://foo",
+            session=session,
+            metrics=pretend.stub(),
+            public_keys_cache=pretend.stub(),
+        )
+        with pytest.raises(osv.OSVPublicKeyAPIError) as exc:
+            vuln_report_verifier.retrieve_public_key_payload()
+
+        assert str(exc.value) == "Invalid response code 418: I'm a teapot"
+        assert exc.value.reason == "public_key_api.status.418"
+
+    def test_retrieve_public_key_payload_json_error(self):
+        response = pretend.stub(
+            text="Still a non-json teapot",
```

Comment:
(Is it bad that I laughed at this given this is a copy of a joke I made :D ) (👍)(🍵)